Index: u-boot-1.2.0/include/configs/gumstix.h
===================================================================
--- u-boot-1.2.0.orig/include/configs/gumstix.h
+++ u-boot-1.2.0/include/configs/gumstix.h
@@ -94,6 +94,9 @@
 				)
 
 #define CONFIG_SUPPORT_VFAT
+#define CONFIG_NETCONSOLE
+#define CONFIG_NET_MULTI
+
 #define CFG_DEVICE_NULLDEV
 
 #include <cmd_confdefs.h>
@@ -162,10 +165,13 @@
 /*
  *  * SMSC91C111 Network Card
  *   */
-#define CONFIG_DRIVER_SMC911X          1
-#define CONFIG_SMC911X_BASE            0x04000000 /* chip select 1         */
+#define CONFIG_DRIVER_SMC91111          1
+#define CONFIG_SMC91111_BASE            0x04000300 /* chip select 1         */
+#undef  CONFIG_SMC_USE_32_BIT                      /* 16 bit bus access     */
+#undef  CONFIG_SMC_91111_EXT_PHY                   /* we use internal phy   */
+#undef  CONFIG_SHOW_ACTIVITY
 #define CONFIG_NET_RETRY_COUNT          10         /* # of retries          */
-#define CONFIG_ETHPRIME		"SMC911X"
+#define CONFIG_ETHPRIME		"SMC91C1111-0"
 
 /*
  * CF slot
Index: u-boot-1.2.0/board/gumstix/memsetup.S
===================================================================
--- u-boot-1.2.0.orig/board/gumstix/memsetup.S
+++ u-boot-1.2.0/board/gumstix/memsetup.S
@@ -394,7 +394,10 @@ initirqs:
 	beq	60b
 #endif
 
-	/* Exercise ethernet reset line - nRESET was cleared when GPIO set up up above */
+	/* Exercise ethernet reset line - the reset on SMC91C111 is opposite polarity to SMC911X! */
+	ldr		r0,	=GPSR3
+	ldr		r1,	=0x00000800
+	str		r1,	[r0]
 
 	ldr	r3, =OSCR		/* reset the OS Timer Count to zero */
 	mov	r2, #0 
@@ -405,7 +408,7 @@ initirqs:
 	cmp	r4, r2
 	bgt	70b 
 
-	ldr		r0,	=GPSR3
+	ldr		r0,	=GPCR3
 	ldr		r1,	=0x00000800
 	str		r1,	[r0]
 
