Index: u-boot-1.2.0/Makefile
===================================================================
--- u-boot-1.2.0.orig/Makefile
+++ u-boot-1.2.0/Makefile
@@ -215,7 +215,7 @@ LIBS := $(addprefix $(obj),$(LIBS))
 .PHONY : $(LIBS)
 
 # Add GCC lib
-PLATFORM_LIBS += -L $(shell dirname `$(CC) $(CFLAGS) -print-libgcc-file-name`) -lgcc
+PLATFORM_LIBS += -L $(shell dirname `$(CC) $(CFLAGS) -print-libgcc-file-name`) -lgcc lib_arm/libarm.a
 
 # The "tools" are needed early, so put this first
 # Don't include stuff already done in $(LIBS)
Index: u-boot-1.2.0/cpu/pxa/config.mk
===================================================================
--- u-boot-1.2.0.orig/cpu/pxa/config.mk
+++ u-boot-1.2.0/cpu/pxa/config.mk
@@ -26,11 +26,11 @@ PLATFORM_RELFLAGS += -fno-strict-aliasin
 	-msoft-float
 
 #PLATFORM_CPPFLAGS += -mapcs-32 -march=armv4 -mtune=strongarm1100
-PLATFORM_CPPFLAGS += -march=armv5 -mtune=xscale
+PLATFORM_CPPFLAGS += -pipe -Os -march=armv5te -mtune=xscale -Wa,-mcpu=xscale
 # =========================================================================
 #
 # Supply options according to compiler version
 #
 # ========================================================================
-PLATFORM_CPPFLAGS +=$(call cc-option,-mapcs-32,-mabi=apcs-gnu)
+#PLATFORM_CPPFLAGS +=$(call cc-option,-mapcs-32,-mabi=apcs-gnu)
 PLATFORM_RELFLAGS +=$(call cc-option,-mshort-load-bytes,$(call cc-option,-malignment-traps,))
Index: u-boot-1.2.0/lib_arm/div0.c
===================================================================
--- u-boot-1.2.0.orig/lib_arm/div0.c
+++ u-boot-1.2.0/lib_arm/div0.c
@@ -22,9 +22,9 @@
  */
 
 /* Replacement (=dummy) for GNU/Linux division-by zero handler */
-void __div0 (void)
+void raise(int dummy)
 {
-	extern void hang (void);
+	extern void hang(void);
 
 	hang();
 }
