Index: u-boot-1.2.0/common/cmd_flash.c
===================================================================
--- u-boot-1.2.0.orig/common/cmd_flash.c
+++ u-boot-1.2.0/common/cmd_flash.c
@@ -45,6 +45,11 @@ int find_dev_and_part(const char *id, st
 
 extern flash_info_t flash_info[];	/* info for FLASH chips */
 
+static const unsigned int JFFS2_CLEAN_HEADER[] = {	0x20031985,
+							0x0000000c,
+							0xe41eb0b1
+						};
+
 /*
  * The user interface starts numbering for Flash banks with 1
  * for historical reasons.
@@ -306,6 +311,15 @@ int do_flinfo ( cmd_tbl_t *cmdtp, int fl
 	return 0;
 }
 
+void jffs2_write_header(flash_info_t *info, unsigned int first, unsigned int last)
+{
+	unsigned int i;
+	for(i=first; i<=last; i++)
+	{
+		flash_write(JFFS2_CLEAN_HEADER, info->start[i], 12);
+	}
+}
+
 int do_flerase (cmd_tbl_t *cmdtp, int flag, int argc, char *argv[])
 {
 	flash_info_t *info;
@@ -317,17 +331,25 @@ int do_flerase (cmd_tbl_t *cmdtp, int fl
 	u8 dev_type, dev_num, pnum;
 #endif
 	int rcode = 0;
+	int jffs2erase = 0;
 
 	if (argc < 2) {
 		printf ("Usage:\n%s\n", cmdtp->usage);
 		return 1;
 	}
 
+	// Special JFFS2 erase which will write a JFFS2 "clean" header after the erase
+	if('j' == argv[0][0])
+	{
+		jffs2erase = 1;
+	}
+	
 	if (strcmp(argv[1], "all") == 0) {
 		for (bank=1; bank<=CFG_MAX_FLASH_BANKS; ++bank) {
 			printf ("Erase Flash Bank # %ld ", bank);
 			info = &flash_info[bank-1];
 			rcode = flash_erase (info, 0, info->sector_count-1);
+			if(jffs2erase) jffs2_write_header(info, 0, info->sector_count-1);
 		}
 		return rcode;
 	}
@@ -340,6 +362,7 @@ int do_flerase (cmd_tbl_t *cmdtp, int fl
 		printf ("Erase Flash Sectors %d-%d in Bank # %d ",
 			sect_first, sect_last, (info-flash_info)+1);
 		rcode = flash_erase(info, sect_first, sect_last);
+		if(jffs2erase) jffs2_write_header(info, sect_first, sect_last);
 		return rcode;
 	}
 
@@ -384,6 +407,7 @@ int do_flerase (cmd_tbl_t *cmdtp, int fl
 		printf ("Erase Flash Bank # %ld ", bank);
 		info = &flash_info[bank-1];
 		rcode = flash_erase (info, 0, info->sector_count-1);
+		if(jffs2erase) jffs2_write_header(info, 0, info->sector_count-1);
 		return rcode;
 	}
 
@@ -397,6 +421,12 @@ int do_flerase (cmd_tbl_t *cmdtp, int fl
 		return 1;
 	}
 
+	if (jffs2erase)
+	{
+		printf ("Usage:\n%s\n", cmdtp->usage);
+		return 1;
+	}
+
 	rcode = flash_sect_erase(addr_first, addr_last);
 	return rcode;
 }
@@ -704,6 +734,16 @@ U_BOOT_CMD(
 );
 
 U_BOOT_CMD(
+	jerase,  3,   1,  do_flerase,
+	"jerase  - erase FLASH memory for JFFS2\n",
+	"jerase N:SF[-SL]\n   - erase sectors SF-SL in FLASH bank # N\n"
+	"jerase bank N\n   - erase FLASH bank # N\n"
+	"jerase all\n   - erase all FLASH banks\n"
+	" in each case, after erasing a flash sector, that sector\n"
+	" has a JFFS2 'clean' marker written to it\n"
+);
+
+U_BOOT_CMD(
 	protect,  4,  1,   do_protect,
 	"protect - enable or disable FLASH write protection\n",
 	"on  start end\n"
