Index: u-boot-1.2.0/fs/jffs2/jffs2_1pass.c
===================================================================
--- u-boot-1.2.0.orig/fs/jffs2/jffs2_1pass.c
+++ u-boot-1.2.0/fs/jffs2/jffs2_1pass.c
@@ -1221,6 +1221,10 @@ jffs2_1pass_build_lists(struct part_info
 					printf("OOPS Padding has bad size "
 						"%d < %d\n", node->totlen,
 						sizeof(struct jffs2_unknown_node));
+			} else if (node->nodetype == JFFS2_NODETYPE_ERASEBLOCK_HEADER) {
+				// Could check offset here to ensure we're at start of erase block
+			} else if (node->nodetype == JFFS2_NODETYPE_SUMMARY) {
+				// Ignore summary blocks for now
 			} else {
 				printf("Unknown node type: %x len %d "
 					"offset 0x%x\n", node->nodetype,
Index: u-boot-1.2.0/include/jffs2/jffs2.h
===================================================================
--- u-boot-1.2.0.orig/include/jffs2/jffs2.h
+++ u-boot-1.2.0/include/jffs2/jffs2.h
@@ -90,6 +90,9 @@
 #define JFFS2_NODETYPE_CLEANMARKER (JFFS2_FEATURE_RWCOMPAT_DELETE | JFFS2_NODE_ACCURATE | 3)
 #define JFFS2_NODETYPE_PADDING (JFFS2_FEATURE_RWCOMPAT_DELETE | JFFS2_NODE_ACCURATE | 4)
 
+#define JFFS2_NODETYPE_ERASEBLOCK_HEADER (JFFS2_FEATURE_RWCOMPAT_DELETE | JFFS2_NODE_ACCURATE | 5)
+#define JFFS2_NODETYPE_SUMMARY (JFFS2_FEATURE_RWCOMPAT_DELETE | JFFS2_NODE_ACCURATE | 6)
+
 /* Maybe later... */
 /*#define JFFS2_NODETYPE_CHECKPOINT (JFFS2_FEATURE_RWCOMPAT_DELETE | JFFS2_NODE_ACCURATE | 3) */
 /*#define JFFS2_NODETYPE_OPTIONS (JFFS2_FEATURE_RWCOMPAT_COPY | JFFS2_NODE_ACCURATE | 4) */
