Index: u-boot-1.2.0/common/cmd_load.c
===================================================================
--- u-boot-1.2.0.orig/common/cmd_load.c
+++ u-boot-1.2.0/common/cmd_load.c
@@ -436,10 +436,11 @@ char his_quote;      /* quote chars he'l
 
 int do_load_serial_bin (cmd_tbl_t *cmdtp, int flag, int argc, char *argv[])
 {
+	DECLARE_GLOBAL_DATA_PTR;
 	ulong offset = 0;
 	ulong addr;
 	int load_baudrate, current_baudrate;
-	int rcode = 0;
+	int rcode = 0, i;
 	char *s;
 
 	/* pre-set offset from CFG_LOAD_ADDR */
@@ -455,6 +456,22 @@ int do_load_serial_bin (cmd_tbl_t *cmdtp
 	if (argc >= 2) {
 		offset = simple_strtoul(argv[1], NULL, 16);
 	}
+
+	if(offset > 0x5C000000 && offset < 0x5C040000) goto offset_in_ram; // Offset in SRAM
+
+	for(i=0; i<CONFIG_NR_DRAM_BANKS; i++)
+	{
+		if(gd->bd->bi_dram[i].size)
+			if(offset >= gd->bd->bi_dram[i].start && offset < (gd->bd->bi_dram[i].start + gd->bd->bi_dram[i].size))
+				goto offset_in_ram;
+	}
+
+	// Not inside any RAM bank
+	printf("Cannot load to 0x%08lX -- address not in RAM\n",offset);
+	return 1;
+
+	offset_in_ram:
+
 	if (argc == 3) {
 		load_baudrate = (int)simple_strtoul(argv[2], NULL, 10);
 
