Index: u-boot-1.2.0/cpu/pxa/serial.c
===================================================================
--- u-boot-1.2.0.orig/cpu/pxa/serial.c
+++ u-boot-1.2.0/cpu/pxa/serial.c
@@ -50,6 +50,12 @@ void serial_setbrg (void)
 		quot = 16;
 	else if (gd->baudrate == 115200)
 		quot = 8;
+	else if (gd->baudrate == 230400)
+		quot = 4;
+	else if (gd->baudrate == 460800)
+		quot = 2;
+	else if (gd->baudrate == 921600)
+		quot = 1;
 	else
 		hang ();
 
